elastic/fleet-server
# instrumentation: Add initial APM tracing

## What is the problem this PR solves?

Adds support to enable instrumentation via the APM Go agent. New config
options have been added to the `Server` input which could be set up in
the `fleet-server` integration configuration.

The added instrumentation covers the `fleet-server` http server and the
`go-elasticsearch` client.

A sample of the configuration that's been added (`instrumentation`):

```yaml
inputs:
  - type: fleet-server
    server:
      host: localhost
      port: 8220
      instrumentation:
        enabled: true
        hosts: ["localhost:8200"]
        environment: production
        secret_token: token
        api_key: apikey
```

## How does this PR solve the problem?

By using the `go.elastic.co/apm` client.

## How to test this PR locally


1. Build `elastic-agent` and patch it with a locally built version of `fleet-server`.
2. Run fleet server in agent mode.
3. Go to Kibana and install the APM Integration
4. Modify fleet server and add these custom settings:
```yml
server.instrumentation:
  enabled: true
  environment: test
```

<img width="1740" alt="image" src="https://user-images.githubusercontent.com/7286993/142614989-d06fda2c-9b44-47b2-a8ca-5311704d3a4f.png">

5. Navigate to the APM App and verify that traces are shown and have the following metadata set:

<img width="1740" alt="Screen Shot 2021-11-19 at 7 22 38 PM" src="https://user-images.githubusercontent.com/7286993/142614830-9c4de934-8a85-4c03-8049-2100e2e7abe8.png">
<img width="1740" alt="image" src="https://user-images.githubusercontent.com/7286993/142615357-b781f985-82ab-4cea-9e76-a70ee5303e06.png">

6. Verify `elasticsearch` shows up as a dependency in the APM Service map

<img width="1740" alt="image" src="https://user-images.githubusercontent.com/7286993/142789258-a6f13e06-948a-45df-bc12-cc1ae4956cba.png">


## Checklist



~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## Related issues

Closes #850
